lib/eventbus.js

```diff
@@ -0,0 +1,229 @@
+            // schema URL returned by schemaUrlExtractor, and then return the event.
+            // The event may be modified; e.g. if the schema has a default for a field
+            // but the event doesn't have it set. If the event failed validation,
+            // an EventInvalidError will be thrown.
+            const validEvent = await this.validate(event);
+            // Now produce the validated event.
+            this.log.trace(
+                { event },
+                `${this.eventRepr(event)} passed schema validation, producing...`
+            );
+            const produceResult = await this.produce(validEvent);
+            // all went fine, return a success status
+            return new EventStatus('success', produceResult, validEvent);
+        } catch (err) {
+            if (err instanceof EventInvalidError) {
+                this.log.debug(
+                    { event, err },
+                    `${this.eventRepr(event)} failed schema validation.`
+                );
+                return new EventStatus('invalid', err, event);
+            } else {
+                this.log.error(
+                    { event, err },
+                    `${this.eventRepr(event)} encountered an error: ${err.message}`
+                );
+                return new EventStatus('error', err, event);
+            }
+        }
+    }
+
+    /**
+     * Validates and produces events.
+     * @param {Array<Object>} events
+     * @return {Object<Array>} of event validate/produce status keyed by status type.
+     */
+    process(events) {
+        // call _processEvent for every event.
+        return P.map(events, event => this._processEvent(event))
+        // Then group the array of EventStatus results to by result.status
+        .then((results) => {
+            // Group event validation and production results by status
+            // Some could succeed and some fail.
+            results = _.groupBy(results, result => result.status);
+
+            // Make sure each possible result type is at least an empty array
+            // to avoid undefined errors later. If there are no errors
+            // results.error should not be undefined, it will be an empty array.
+            _.each(EventStatus.STATUSES, (status) => {
+                results[status] = results[status] || [];
+            });
+
+            // Convert any failed events to error events and produce them
+            // if given mapToErrorEvent was configured.
+            // This should be done in the background.
+            const failedResults = results.invalid.concat(results.error);
+            if (this.mapToErrorEvent && !_.isEmpty(failedResults)) {
+
+                // Process each error event to validate and produce it.
+                // TODO should we ignore results like this?  They will be logged...
+                // TODO is setTimeout(..., 0) necessary here?  Not sure.  Could we just
+                // do P.map() to make this background async?
```

Comment:
Ya set timeout is confusing

Reply:
Assuming that your comment means I'm doing this right!  Let me know if that's not the case!